### DISCUSSION TITLE
[Chapter 03] 부트가 아닌 스프링에서의 @ComponentScan의 사용 사례

### DISCUSSION THREAD

**Discussion by JoisFe** [cowaine/spring-boot-for-msa#15] [Chapter 03. Spring Application Basics]
``` java
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Inherited
@SpringBootConfiguration
@EnableAutoConfiguration
@ComponentScan(excludeFilters = { @Filter(type = FilterType.CUSTOM, classes = TypeExcludeFilter.class),
		@Filter(type = FilterType.CUSTOM, classes = AutoConfigurationExcludeFilter.class) })
public @interface SpringBootApplication {
```

스프링 부트를 이용한 프로젝트를 진행하면 `@ComponentScan` 애너테이션을 사용하지 않아도 자바 설정 클래스 or 스테레오 애너테이션이 선언된 클래스들을 잘 스캔하고 있습니다.
그 이유는 애플리케이션 클래스에 있는 `@SpringBootApplication` 에 `@ComponentScan`이 포함되어 있기 때문이고, value 혹은 basePackages가 설정되어 있지 않아도 default로 `@ComponentScan` 이 정의된 클래스가 위치한 패키지가 기본값이 되므로 해당 메인 클래스가 존재하는 패키지 혹은 하위 패키지에 있는 자바 설정 클래스 혹은 스테레오 타입 애너테이션이 선언된 클래스를 스캔하는 것 이었고, 이로 인해 스프링부트를 사용하다 보면 이러한 부분을 크게 신경쓰지 않아도 되는 장점이 있지만, `@ComponentScan` 애너테이션에 대해 잘 알지 못하게되는 문제도 있는 것 같습니다.

부트와 같이 `@SpringBootApplication` 애너테이션이 없는 스프링 경우에는 `@ComponentScan` 애너테이션을 어떻게 사용하는지 경험을 공유해주실 분 계신가요?

**Comment by ahngilwoong** [comment]
![image](https://github.com/cowaine/spring-boot-for-msa/assets/65801290/f90d2af8-5e38-4787-abac-abff20348a80)
모 교육의 예제코드가 생각나서 가져왔습니다.
보통 boot환경이 아닌 Spring에서 Component Scan 설정을 xml, class 두가지 방식으로 할텐데

구분 | XML 방식 설정 | Java 방식 설정 클래스(권장) | Java 설정 방법
| --- | --- | --- | --- |
|DispatcherServlet 설정 | web.xml |SpringConfig.java | AbstractAnnotationConfigDispatcherServletInitializer 상속 또는 WebApplicationInitializer 인터페이스 구현|
|서블릿 영역 설정 | servlet-context.xml | ServletConfig.java | @EnableWebMvc, @Configuration, @ComponentScan 를 사용하여 웹 관련 빈 등록|
|어플리케이션 영역 설정 | root-context.xml | RootConfig.java |  마찬가지로 @Configuration과 @ComponentScan으로 Service와 Repository Component들을 스캔하도록 한다.|

책에서는 Boot는 ApplicationContext를 하나만 두어 관리한다고 적혀있었는데 말씀주신대로 Spring MVC 프로젝트에서는
이렇게 각각의 영역별로 2개의 컴포넌트 스캔을 두어 필요한 빈들의 BasePackage를 정해 웹 관련 설정, 그외 전체 공유 빈 등을 따로 관리를 한 경험이 생각나 공유합니다.
책에 나온 Groovy로 설정하는 방법에 대해서도 알아가보고싶습니다.
